### DISCUSSION TITLE
Hydrostatic stress coupled with concentration

### DISCUSSION THREAD

**Discussion by sidharthsarmah** [idaholab/moose#22910] [Q&A General] [answered]
Hi @hugary1995 ,

I want to implement this PDE term of the diffusion equation in a kernel, the CL is the variable to be computed which is coupled and dependent on the hydrostatic stress gradient.

Could you suggest a similar example of how I proceed, the CL is the variable and \sigma_h is the aux variable.

**Comment by hugary1995** [comment, marked as answer]
Yeah sure. Are you using AD?

**Reply by hugary1995** [reply]
You'll have to tell me how to compute sigma_h in your third term. Be it prescribed, or related to another nonlinear variable, please write down the equation.

**Reply by hugary1995** [reply]
Okay, that was my guess. So unfortunately this is not a good approach. Let me explain.

The whole purpose of using AD is to get the exact algorithmic tangent, so that you get quadratic convergence using a Newton solver. MOOSE uses forward mode automatic differentiation to get the algorithmic tangent (a.k.a. Jacobian). All you need to know as a user is that all the operations carries derivatives with them, so that the derivative information can be "propagated" down to the final residual. For this to work, all the operations you use during the calculation of residual have to support AD.

However, that is not the case here since you use an AuxVariable to compute the hydrostatic stress (or the gradient of hydrostatic stress). An AuxVariable does not carry derivative information with it. As a result, although your final residual will still be correct (if implemented correctly), the algorithmic tangent will be inexact. Then it essentially loses the entire point of using AD.

There are multiple ways of dealing with this. For a new user, the easiest approach is probably to reparameterize the equation by projecting the hydrostatic stress onto an additional nonlinear variable, let's call it p_h. You need to solve for an additional equation for p_h:
w * (p_h - sigma_h) = 0
where w is the test function, p_h is the nonlinear variable to be solved, and sigma_h is the hydrostatic stress computed at quadrature points (as an AD material property).

The next step is to replace sigma_h in the third term of your original equation by this new variable p_h.

**Reply by hugary1995** [reply]
I suggest you first figure out how to compute a scalar-valued hydrostatic stress from a tensor-valued stress as an AD material property. There should a material object in the tensor_mechanics module that allows you to do that. If you don't find it, I can track that down for you when I get off work.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995,

I am a bit confused on this part.

The hydrostatic stress can directly be declared this way below since its defined as a RankTwoScalar parameter already (https://mooseframework.inl.gov/source/utils/RankTwoScalarTools.html).

Do I have to use this auxvariable and convert it to a ADMaterialProperty and then relate it to p_h
```
[AuxVariables]

  [./hydrostatic]
    order = CONSTANT
    family = MONOMIAL
  [../]

[AuxKernels]


OR

I have to defined another ADMaterialProperty named hydrostatic_stress = sum of normal stresses and then relate it to p_h?

**Reply by hugary1995** [reply]
No, let me reiterate, don't use aux variable to bridge nonlinear equations in AD.

You need to create an AD material property for hydrostatic stress using https://mooseframework.inl.gov/source/materials/RankTwoInvariant.html

**Reply by sidharthsarmah** [reply]
Hi @hugary1995, thanks!

Please suggest:

I am a bit confused on how to relate pressure with hydrostatic_stress as you mentioned. 

Based on the notes, "_This Material model is set up by [TensorMechanics/Master] automatically when stress components are requested in the generate_output parameter, but can also be set up directly by the user._"

I have declared a variable named pressure and hydrostatic_stress directly by calling generate_output:
```
[Variables]

  [./conc_lattice]
  initial_condition = 1.0e3
  [../]

  [./pressure]
  [../]

[]
```
```
[Modules/TensorMechanics/Master]

  [./all]
    strain = SMALL
    incremental = true
    add_variables = true
    generate_output = 'stress_xx stress_yy stress_zz hydrostatic_stress'
  use_automatic_differentiation = true
  [../]

[]
```
Now since conc_lattice (C_L) is the primary kernel to be solved in this equation. Can we convert this convert this equation as follows with grad of hydrostatic stress in the coefficient?


Can you suggest on how to:
1. Relate the variable pressure to the property hydrostatic_stress?
2. How do I extract the grad of hydrostatic_stress in a Kernel?
```
[Kernels]

  [./stress_grad]
    type = ADMatDiffusion
    variable = 'conc_lattice'
    diffusivity = 'func2'
  [../]

[]
```
```
[Materials]

  [./constants]
    type = ADGenericConstantMaterial
    prop_names  = 'DL		NT		NL		WB	VH	R	T'
    prop_values = '0.01271	8.464e13	8.464e17	11.5e6	2000.0	8314.5	300'
  [../]


  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 201880
    poissons_ratio = 0.3
  [../]

  [./stress]
    type = ADComputeLinearElasticStress
  [../]

[]
```
3. Kernel which can be used?

**Reply by hugary1995** [reply]
We should tackle this problem step by step, and please try to follow my suggestions -- I am trying to help.

First, please use the object I referenced above to obtain the hydrostatic stress sigma_h as an AD material property. Please convince yourself it is working as expected.

Next, as I said earlier, to relate p_h and sigma_h, you need to a projection by solving the following equation:
w (p_h - sigma_h) = 0

Finally, with  nonlinear variable p_h, it will be trivial to obtain he pressure gradient.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995, please suggest if I am correct.

Step 1:
_"First, please use the object I referenced above to obtain the hydrostatic stress sigma_h as an AD material property."_

Based on the object, "_This Material model is set up by [TensorMechanics/Master] automatically when stress components are requested in the generate_output parameter, but can also be set up directly by the user._"

The hydrostatic_stress is set up as an ADProperty by either of the two ways:

1. Directly by calling generate_output(https://mooseframework.inl.gov/source/materials/RankTwoInvariant.html), is it correct?
```
[Modules/TensorMechanics/Master]
  [./all]
          strain = SMALL
          incremental = true
          add_variables = true
          generate_output = 'hydrostatic_stress'
          use_automatic_differentiation = true
  [../]
[]
```
2. Invoked as a material property on calling the material object, 'ADComputeLinearElasticStress':

```
[Materials]

***Elastic case***
  [./elas_stress]
    type = ADComputeLinearElasticStress
  [../]

[]
```

Step 2:
_"Next, as I said earlier, to relate p_h and sigma_h, you need to a projection by solving the following equation:
w (p_h - sigma_h) = 0"_

Now the 'hydrostatic_stress' property is assigned to a non-linear variable 'pressure' by using the kernel, "ADMaterialPropertyValue"(https://mooseframework.inl.gov/source/kernels/MaterialPropertyValue.html):


Step 3:
Please let me know if the above steps are correct, I can then work on to find the final term, thanks!

_'Finally, with nonlinear variable p_h, it will be trivial to obtain he pressure gradient.'_

**Reply by hugary1995** [reply]
The action only outputs hydrostatic stress as an AuxVariable. Again, you need to use the object I referenced above, which is called `RankTwoInvariant`. In your case you need to use its AD version called `ADRankTwoInvariant`. First, set up this object, run the model, output the material property, and verify that the value is correct.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995, thanks

I faced an issue implementing it as as no examples were available, my bad!

I tried implementing the object now, but found an error, please suggest!

```
*** ERROR ***
The following error occurred in the object "stress_pressure", of type "ADRankTwoInvariant".

Not a recognized invariant for RankTwoInvariant
```
Object:
```
[Materials]


  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 201880
    poissons_ratio = 0.3
  [../]

  [./elas_stress]
    type = ADComputeLinearElasticStress
  [../]

[]
```

**Reply by hugary1995** [reply]
Glad you are making progress! The error message is helpful. If you go to the documentation page about this object: https://mooseframework.inl.gov/source/materials/RankTwoInvariant.html
Under section "Optional Parameters", there is a parameter called "invariant". You need to choose an invariant from the available options. In your case, `invariant = Hydrostatic`.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995,
I included it:
```
[Materials]


  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 201880
    poissons_ratio = 0.3
  [../]

  [./elas_stress]
    type = ADComputeLinearElasticStress
  [../]

[]
```
I was able to solve it, thanks!
```
2 Nonlinear |R| = 2.605953e-12
 Solve Converged!
Finished Executing                                                                       [ 29.06 s] [   18 MB]
```
With step 1, being done. Step 2 is to assign this property to a nonlinear variable, say 'pressure'!

Is it this the way to do it?

```
  [./ph_stress]
            type = ADMaterialPropertyValue
            prop_name = 'hydrostatic_stress'
            variable = 'pressure'
  [../]
```

**Reply by hugary1995** [reply]
Yep. Again, after it runs, look at the exodus output and make sure the result makes sense.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995 , thanks for this, I was able to obtain the .exo file!

I am planning to generate the results based on a article.

Now for step 3:



The way I plan to do this is; since the first and third term are identical , I am planning to use the kernel "ADMatDiffusion" as follows:

<img width="245" alt="image" src="https://user-images.githubusercontent.com/110927986/207140609-8086b376-e7c5-4cff-a082-6fa2f633c123.png">

Do you think this is correct? As both CL and pressure are non linear variable, I can use either of them as gradient of the primary variable of the Kernel!

Kernel for term 1:
```
[Kernels]

  [./diffusion_grad]
    type = ADMatDiffusion
    variable = 'conc_lattice'
    diffusivity = '-DL' #(- sign, as it is there in the term 1)
  [../]

[]
```
Kernel for term 3:
```
[Kernels]



[]

[Materials]

  [./elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 211.0e9
    poissons_ratio = 0.3
  [../]

  [./elas_stress]
    type = ADComputeLinearElasticStress
  [../]

[]

**Reply by hugary1995** [reply]
Not quite, you're almost there. When there are multiple kernels associated with multiple nonlinear variables, MOOSE groups kernels  into equations using the `variable` parameter in each Kernel. For example
```python
  [diffusion_grad]
    type = ADMatDiffusion
    variable = 'conc_lattice'
    diffusivity = '-DL' #(- sign, as it is there in the term 1)
  []
```
associates this term with the "conc_lattice" equation, which is correct. However, 
```python
  [stress_grad]
    type = ADMatDiffusion
    variable = 'pressure'
    diffusivity = 'func2'
  []
```
will associate this term with the "pressure" equation. That's not you want, as this is the third term of your "conc_lattice" equation. You need to specify "pressure" as a coupled variable instead:
```python
  [stress_grad]
    type = ADMatDiffusion
    variable = 'conc_lattice'
    v = 'pressure'
    diffusivity = 'func2'
  []
```

Otherwise, everything looks good to me.

**Reply by sidharthsarmah** [reply]
Hi @hugary1995 thanks,

This is something what I thought as well, but my doubt in this is, just for curiosity:

The form of ADMatDiffusion is as follows:
<img width="416" alt="image" src="https://user-images.githubusercontent.com/110927986/207161385-4d1073f2-4665-4732-bf38-723962223a72.png">


I want the y to be the hydrostatic_stress and x contains the CL.

How will MOOSE decides when you put a 'variable = CL' and couple 'v = pressure' to be x and y respectively? Because based on the ADMatDiffusion kernel, the primary 'variable = CL' goes into the gradient inside? Am I missing something?

Basically I want to know how will MOOSE decides whether the primary variable or couple variable should represent ```y```, which is basically the gradient of the variable for kernel ADMatdiffusion

<img width="287" alt="image" src="https://user-images.githubusercontent.com/110927986/207163906-a284b2a2-5525-45de-8291-d29d9536ce3e.png">

**Reply by hugary1995** [reply]
I see your confusion. There is a if-statement in the source checking whether `v` is coupled. If `v` is found to be coupled, then it will use the gradient of `v`. if not, it will be the gradient of `variable`. The documentation also explains this pretty well: https://mooseframework.inl.gov/moose/source/kernels/ADMatDiffusion.html

**Reply by sidharthsarmah** [reply]
Hi @hugary1995, thanks for the clarification, so if I introduce a couple ```v```, then with no change in the .h and .C files, the gradient will act on the couple ```v``` variable, unless I can the if condition to work on the ```variable```!

That's sorted, I shall once implement all the terms using kernels and show it to you.

Th next thing I want to do is implement the constitutive stress-strain law, can you suggest how do I approach?

<img width="821" alt="image" src="https://user-images.githubusercontent.com/110927986/207360369-4025340b-5cb8-4962-b1b5-3a7a3052fe06.png">

**Reply by hugary1995** [reply]
Could you open a new discussion so that it's easier to search in the future?

**Reply by sidharthsarmah** [reply]
Hi @hugary1995, its done, please suggest, thanks!
https://github.com/idaholab/moose/discussions/22957
